### PR TITLE
Support custom status messages

### PR DESCRIPTION
**Pull request by turt2live**
Part of https://github.com/vector-im/riot-web/issues/1528

Status messages are short strings of text that users can set to be visible to other users. The js-sdk bits here are for setting and reading the status message for a given user.

Status messages are completely unrelated to the presence status message in that they do not expire and have a different intent. Presence status messages are intended to be read as "Online reading documentation" whereas status messages would be just "Reading documentation" (no associative presence). 

This currently uses a state event in direct chats to represent the status. It is intended that in the future there will be extensible profiles which can better communicate this information. Because this is a feature we're trying out, the event type is `im.vector.user_status` instead of something in the Matrix namespace. The API here should be abstract enough that changing to extensible profiles in the future should be easy.

----

**`im.vector.user_status` spec**:

The user status event is a state event with a state key of the user ID the status is for with a single `status` field in the content. The `status` field is a string of preferably limited length (<60 characters, not enforced) where falsey values (empty, null, undefined, not set, etc) equate to "no status" whereas a string with content represents the user's status. 

It is not intended at this time to support per-room status messages, despite using a state event. This is why this js-sdk PR happily overwrites the user's status message when it sees an `im.vector.user_status` state event. Simply seeing the state event in any room will change that user's status message everywhere.

### CODE REVIEW

**Comment by Half-Shot** [comment]
I have reservations about including more `im.*` specific unspecced things into the js-sdk :/

**Comment by turt2live** [comment]
The event itself should be fairly well abstracted so it can be replaced with extensible profiles in the future, as mentioned in the description. This is intentionally being done without a spec proposal (hence the vector namespace) to get the feature in front of people and iterate upon it while extensible profiles are being worked on.

**Comment by Half-Shot** [comment]
To clarify, my worry is people will see `setStatusMessage` and believe it is the defacto way to set a status message in Matrix, while this is a im.vector brand feature. I don't really have an issue with this ending up in the spec as a feature. Anyway, we sort of discussed this oob and I expect everyone is aware of this already.

**Comment by turt2live** [comment]
@Half-Shot I've made the entire API be prefixed with `_unstable_` in https://github.com/matrix-org/matrix-js-sdk/pull/805/commits/08b3dfa3b5b5d0b63272f0b80b9fdd88d0795c45 - does this alleviate your concerns? Also, if you have suggestions for making it not as ugly please let me know.

File: src/models/user.js [hunk 1]
```diff
@@ -39,13 +39,17 @@ limitations under the License.
  *                when a user was last active.
  * @prop {Boolean} currentlyActive Whether we should consider lastActiveAgo to be
  *               an approximation and that the user should be seen as active 'now'
+ * @prop {string} statusMessage The status message for the user, if known. This is
```
Comment (bwindels) [inline src/models/user.js]: `_unstable_statusMessage` for now?

**Comment by Half-Shot** [comment]
>@Half-Shot I've made the entire API be prefixed with _unstable_ in 08b3dfa - does this alleviate your concerns? Also, if you have suggestions for making it not as ugly please let me know.

Thank you :). I don't really know how we could make this better, than perhaps `unstableGetBlah...`

**Comment by turt2live** [comment]
tbh I kinda like `_unstable_` because it makes searching for this stuff a little easier, and having slept on it it doesn't look as bad as it did last night.

**Pull request merged**